app/src/crypto.c

```diff
@@ -584,3 +612,288 @@ zxerr_t crypto_fillMASP(uint8_t *buffer, uint16_t bufferLen, uint16_t *cmdRespon
+
+    return parser_ok;
+}
+static zxerr_t sign_sapling_spend(keys_t *keys, uint8_t alpha[static KEY_LENGTH], uint8_t sign_hash[static KEY_LENGTH], uint8_t *signature) {
+    if (alpha == NULL || sign_hash == NULL || signature == NULL) {
+        return zxerr_no_data;
+    }
+
+    uint8_t data_to_be_signed[2 * HASH_LEN] = {0};
+    uint8_t rsk[KEY_LENGTH] = {0};
+    uint8_t rk[KEY_LENGTH] = {0};
+
+    // get randomized secret
+    randomized_secret_from_seed(keys->ask, alpha, rsk);
+
+    //rsk to rk
+    scalar_multiplication(rsk, SpendingKeyGenerator, rk);
+
+    // sign
+    MEMCPY(data_to_be_signed, rk, KEY_LENGTH);
+    MEMCPY(data_to_be_signed + KEY_LENGTH, sign_hash, HASH_LEN);
+
+    // Get rng
+    uint8_t rng[RNG_LEN] = {0};
+    cx_rng_no_throw(rng, RNG_LEN);
+
+    // Compute r and rbar
+    uint8_t r[32] = {0};
+    uint8_t rbar[32] = {0};
+    CHECK_PARSER_OK(h_star(rng, sizeof(rng), data_to_be_signed, sizeof(data_to_be_signed), r));
+    CHECK_PARSER_OK(scalar_multiplication(r, SpendingKeyGenerator, rbar));
+
+    //compute s and sbar
+    uint8_t s[32] = {0};
+    uint8_t sbar[32] = {0};
+    CHECK_PARSER_OK(h_star(rbar, sizeof(rbar), data_to_be_signed, sizeof(data_to_be_signed), s));
+    CHECK_PARSER_OK(compute_sbar(s, r, rsk, sbar));
+
+    MEMCPY(signature, rbar, HASH_LEN);
+    MEMCPY(signature + HASH_LEN, sbar, HASH_LEN);
+
+    return zxerr_ok;
+}
+
+zxerr_t crypto_sign_spends_sapling(const parser_tx_t *txObj, keys_t *keys, uint8_t *output, uint16_t outputLen, uint16_t *responseLen) {
+    zemu_log_stack("crypto_signspends_sapling");
+    if (txObj->transaction.sections.maspTx.data.sapling_bundle.n_shielded_spends == 0) {
+        return zxerr_ok;
+    }
+
+    MEMZERO(output, outputLen);
+    // Get Signature hash
+    uint8_t sign_hash[HASH_LEN] = {0};
+    signature_hash(txObj, sign_hash);
+
+    uint8_t signature[2 * HASH_LEN] = {0};
+    uint8_t alpha[KEY_LENGTH] = {0};
+    const uint8_t *spend = txObj->transaction.sections.maspBuilder.builder.sapling_builder.spends.ptr;
+    uint16_t spendLen = 0;
+
+    for (uint64_t i = 0; i < txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends; i++) {
+        // Get spend description and alpha
+        spend += spendLen;
+        MEMCPY(alpha, spend + ALPHA_OFFSET, KEY_LENGTH);
+
+        CHECK_ZXERR(sign_sapling_spend(keys, alpha, sign_hash, signature));
+
+        // Copy signature to output
+        MEMCPY(output + i * MASP_SIG_LEN, signature, MASP_SIG_LEN);
+
+        // Get this spend lenght to get next one
+        getSpendDescriptionLen(spend, &spendLen);
+
+        *responseLen += MASP_SIG_LEN;
+    }
+
+    return zxerr_ok;
+}
+
+
+zxerr_t checkSpends(const parser_tx_t *txObj, keys_t *keys) {
+    uint8_t *builder_spend = (uint8_t *)txObj->transaction.sections.maspBuilder.builder.sapling_builder.spends.ptr;
+    uint8_t *tx_spend = (uint8_t *)txObj->transaction.sections.maspTx.data.sapling_bundle.shielded_spends.ptr;
```

Comment:
Add check for `txObj->transaction.sections.maspTx.data.sapling_bundle.shielded_spends.len == txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends * SHIELDED_SPENDS_LEN`
Similar for `builder_spend`, before increasing the pointer, you should check that there are bytes in the buffer